Fix null reference when no env vars defined in yml

davidsiaw opened the pull request:
This PR fixes a bug introduced in #23 where if you do not define a `run_env` section in the `barcelona.yml` it would throw a null reference error.

## How to test
1. Create a `barcelona.yml` containing this:
```
environments:
  hello:
    name: hello
    image_name: quay.io/dsiawdegica/hello
    scheduled_tasks: []
    services:
      - name: web
        service_type: web
        cpu: null
        memory: 256
        command: sh /start.sh
        listeners:
          - endpoint: hello
            health_check_path: /
        port_mappings:
          - lb_port: 80
            container_port: 8080
            protocol: http
```
2.Run `./bcn -d run -e hello -E FOO=bar` ew
3. Observe that there is no segfault error, and it outputs 

```
{"command":"ew","env_vars":{"FOO":"bar"},"interactive":true}
```

degicat commented:
@Resonious please review this

Resonious commented on cmd/run.go:
```diff
@@ -160,6 +153,21 @@ var RunCommand = cli.Command{
+		if err != nil {
+			return nil, err
+		}
+		heritageName = env.Name
```
I'm no Go expert, but this looks to me like we're writing a value to the `heritageName` argument, and never reading it inside of this function. But Go has no pass-by-reference, so the assignment does nothing, right?

davidsiaw replied:
You are right. It is a totally useless line. I'll get rid of it

Resonious commented:
@davidsiaw I downloaded the binary from the CI build job https://gitlab.com/degica/barcelona-cli/-/jobs/504621079 and it still seems to segfault after performing the test instructions.

```
[nigel@resonious testsz]$ cat barcelona.yml 
environments:
  hello:
    name: hello
    image_name: quay.io/dsiawdegica/hello
    scheduled_tasks: []
    services:
      - name: web
        service_type: web
        cpu: null
        memory: 256
        command: sh /start.sh
        listeners:
          - endpoint: hello
            health_check_path: /
        port_mappings:
          - lb_port: 80
            container_port: 8080
            protocol: http

[nigel@resonious testsz]$ ./bcn -d run -e hello -E FOO=bar ew
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x776392]

goroutine 1 [running]:
github.com/degica/barcelona-cli/cmd.loadEnvVars(0x7ffc2466a771, 0x5, 0x0, 0x0, 0xc000115b90, 0x68, 0x1)
        /go/src/github.com/degica/barcelona-cli/cmd/run.go:163 +0xa2
github.com/degica/barcelona-cli/cmd.glob..func38(0xc000136b40, 0x100, 0xc000136b40)
        /go/src/github.com/degica/barcelona-cli/cmd/run.go:52 +0x1ba
github.com/degica/barcelona-cli/vendor/github.com/urfave/cli.HandleAction(0x7c6d20, 0x85f708, 0xc000136b40, 0xc000026600, 0x0)
        /go/src/github.com/degica/barcelona-cli/vendor/github.com/urfave/cli/app.go:485 +0xc8
github.com/degica/barcelona-cli/vendor/github.com/urfave/cli.Command.Run(0x842001, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x851c3b, 0x28, 0x0, ...)
        /go/src/github.com/degica/barcelona-cli/vendor/github.com/urfave/cli/command.go:193 +0x97c
github.com/degica/barcelona-cli/vendor/github.com/urfave/cli.(*App).Run(0xc00010b6c0, 0xc000012080, 0x8, 0x8, 0x0, 0x0)
        /go/src/github.com/degica/barcelona-cli/vendor/github.com/urfave/cli/app.go:250 +0x82a
main.main()
        /go/src/github.com/degica/barcelona-cli/main.go:41 +0x445
[nigel@resonious testsz]$
```

Let me know if there's something I'm missing.

davidsiaw commented:
@Resonious  That is embarassing. I must have messed up. I made a fix.

Resonious commented on cmd/run.go:
```diff
@@ -49,20 +49,13 @@ var RunCommand = cli.Command{
-			if err != nil {
-				return cli.NewExitError(err.Error(), 1)
-			}
-			heritageName = env.Name
```
Looks to me like we might still need this `heritageName = env.Name` part.

Later in the file, it does this
```go
api.DefaultClient.Post("/heritages/"+heritageName+"/oneoffs", bytes.NewBuffer(j))
```
which might not won't work right if `len(envName) > 0 && len(heritageName) == 0`.

Does that sound right?

davidsiaw replied:
Hmm that code looks awfully suspicious. You are right. I'll try and fix it in a sane way

davidsiaw commented:
@Resonious sorry it took me forever to get back to this. I legitimately thought that this was solved but it wasn't. I made a change to fix that issue. I will make an issue to refactor this into a more understandable thing later.

davidsiaw commented:
@Resonious I have rebased the branch against master and fixed the ci, if you have time could you please have another look? I made an issue to fix this messiness in run.go #29

davidsiaw commented:
@Resonious made some fixes to this long-overdue pr. Could you have one last look? Thanks!

Resonious reviewed (approved):
LGTM :+1:
